pymtl3/dsl/test/set_param_test.py

```diff
@@ -0,0 +1,118 @@
+"""
+========================================================================
+set_param_test.py
+========================================================================
+
+Author : Yanhui Ou, Cheng Tan
+Date   : May 28, 2019
+"""
+from __future__ import absolute_import, division, print_function
+
+from pymtl3.dsl.NamedObject import NamedObject
+
+
+class Animal( NamedObject ):
+  def construct( s, lunch="dirt", dinner="dirt" ):
+    s.lunch  = lunch 
+    s.dinner = dinner
+
+class AnimalHouse( NamedObject ):
+  def construct( s, AnimalTypes=[] ):
+    s.animals = [ A() for A in AnimalTypes ]
+
+class HoneyBadger( Animal ):
+  pass
+
+class Dromaius( Animal ):
+  pass
+
+class Panda( Animal ):
+  pass
+
+class Tiger( Animal ):
+  def construct( s, MeatTypes=[], lunch="dirt", dinner="dirt" ):
+    s.meats  = [ Meat() for Meat in MeatTypes ]
+    s.lunch  = lunch
+    s.dinner = dinner
+
+class BabyTiger( Tiger ):
+  pass
+
+class Aquarium( AnimalHouse ):
+  pass
+
+class TigerTerrace( AnimalHouse ):
+  pass
+
+class PandaHouse( AnimalHouse ):
+  pass
+
+class Zoo( NamedObject ):
+  def construct( s, AnimalHouseTypes=[] ):
+    s.houses = [ H() for H in AnimalHouseTypes ]
+
+def test_simple():
+  A = Dromaius()
+  A.set_param( "top.construct", lunch="grass" )
+  A.elaborate()
+
+  print( A.lunch )
+  assert A.lunch  == "grass" 
+  assert A.dinner == "dirt" 
+
+def test_set_param_overwrite():
+
+  Z = Zoo()
+  Z.set_param( "top.construct", AnimalHouseTypes=[ PandaHouse, TigerTerrace, AnimalHouse, Aquarium ] )
+  Z.set_param( "top.houses[0].construct", AnimalTypes=[ Panda, Panda, Panda   ] )
+  Z.set_param( "top.houses[1].construct", AnimalTypes=[ Tiger, BabyTiger      ] )
+  Z.set_param( "top.houses[2].construct", AnimalTypes=[ Dromaius, HoneyBadger ] )
+  Z.set_param( "top.houses[1].animals[0].construct", MeatTypes=[ Dromaius, HoneyBadger ] )
+  Z.set_param( "top.houses[0].animals*.construct", lunch="bamboo", dinner="bamboo" )
+  Z.set_param( "top.houses[1].animals[0].meats*.construct", lunch="bamboo" )
+  Z.set_param( "top.houses[1].animals[0].meats[1].construct", lunch="grass" )
+  # The following set_param test tricky overwrite behavior.
```

Comment:
This breaks the algorithm we discussed yesterday @jsn1993

Reply:
line 75~77

Reply:
hmm why? the reverse orderof ordededdict still works for each parameter

Reply:
hmmm let me think about it. When there are multiple parameters it's not only the name but also the argument name that makes a difference.

Reply:
yes exactly... for line 76, if I delete animals[0] and re-append the new dict, the 'dinner' would be lost
If i pop the node, merge it with the new dict, then re-append it to the end, that still won't work because it screws the order, since that means line 74 may overwrite line 75.

Reply:
In my current algorithm, when inserting a node with '*', it will updates all previous nodes who matches the regex.
It's definitely slower though. I'm trying to do some microbenchmarking to see how much it may slow down the elaboration process.